node/src/Pos/Update/Worker.hs

```diff
@@ -28,12 +35,42 @@ usWorkers =
         recoveryCommGuard $ do
             logDebug "Updating slot for US..."
             processNewSlot s
-            void (fork checkForUpdate)
+            checkForUpdate
 
-checkForUpdate :: WorkMode ssc ctx m => m ()
+checkForUpdate ::
+       forall ssc ctx m. WorkMode ssc ctx m
+    => m ()
 checkForUpdate = do
     logDebug "Checking for update..."
-    proposals <- getConfirmedProposals (Just $ svNumber curSoftwareVersion)
-    logDebug $ sformat ("Potentially relevant proposals: "%listJson)
-                    (cpsUpdateProposal <$> proposals)
-    mapM_ downloadUpdate proposals
+    confirmedProposals <-
+        getConfirmedProposals (Just $ svNumber curSoftwareVersion)
+    case nonEmpty confirmedProposals of
+        Nothing ->
+            logDebug
+                "There are no new confirmed update proposals for our application"
+        Just confirmedProposalsNE -> processProposals confirmedProposalsNE
+  where
+    processProposals :: NonEmpty ConfirmedProposalState -> m ()
+    processProposals confirmedProposals = do
+        let cpsToNumericVersion =
+                svNumber . upSoftwareVersion . cpsUpdateProposal
+        let newestCPS =
+                maximumBy (comparing cpsToNumericVersion) confirmedProposals
+        logInfo $
+            sformat
+                ("There are new confirmed update proposals for our application: "
+                 %listJsonIndent 2%
+                 "\n The newest one is: "%build%" and we want to download it")
+                (cpsUpdateProposal <$> confirmedProposals)
+                (cpsUpdateProposal newestCPS)
+        downloadUpdate newestCPS
```

Comment:
Checks like `isVersionAppropriate` and for `This update is not for our system` should be done here, before picking up maximum

Reply:
I think it doesn't matter, because it's only a sanity check which must always pass (by contract of `getConfirmedProposals`).
There advantages and disadvantages of doing it earlier and later.
If we do it earlier, we detect error in code earlier.
If we do it later, we cover more logic and will detect if something is broken somewhere in between this function and the place where we check it.
In theory we can do this check in 100500 places or can never do it. I don't think it's essential enough to really care about it.

Reply:
@gromakovsky but what will happen if there are proposal A which has installer for Windows and proposal B for MacOS. And B hash highest version than A. And then widnows can't be updated to A, though it seems beneficial and good thing to do

Reply:
Ah, you meant different check, ok, got it.

Reply:
@georgeee ok, I changed `getAllConfirmedProposals` so that it returns only those proposals which have data for our `SystemTag`.